### ISSUE TITLE
[resize-observer] ResizeObserver should support fragments

### ISSUE THREAD

**Issue by gregwhitworth** [w3c/csswg-drafts#3673]
@plinss Brought up the desire for the API to understand the notion of fragments in the TAG review (https://github.com/w3ctag/design-reviews/issues/187#issuecomment-362633185). I've decided to make this a seperate issue because I would like to understand this a bit further. I'm assuming that you wouldn't expect to be able to attach an observer to a fragment correct (since we won't know how many fragments will exist in a given layout until after layout, thus hard coding the fragment doesn't make much since)?

So my suspicion is that you're wanting the fragments represented in the resulting entry object, is this correct?

**Comment by plinss** [comment]
Correct, I haven't spent any time trying to design a specific api surface for this, but anything that exposes all the fragments in the entry rather than presuming one box is what I'm trying to get.

**Comment by gregwhitworth** [comment]
Thanks - that clarifies it for me. So effectively that would require a change to the specification in two ways:

**1. The entry object**

This would need to have a notion of fragments although I don't desire to denote that in any way. Taking a scenario such as multi-column - we could adjust the `ResizeObserverEntry` to a <boxName>: sequence<ResizeObserverSize> which would end up with a result of something like this, again assuming multi-col with two resulting columns and the author observing the contentBox:

```
contentBox: [
        {
           inlineSize: <double>,
           blockSize: <double>
        },
        {
           inlineSize: <double>,
           blockSize: <double>
        }
]
```

This has the added benefit, to @atotic point in the TAG review of the most common use cases currently not being about fragmentation so even if an implementation can't initially support building out the fragments easily or performantly they can add it as demand for the capabilty increases and we don't have to revisit the API or come up with a brand new one to support it.

**2. Observing numerous boxes**
I'll need to thing about this a bit more thoroughly to understand if there is anything that currently needs to change with the how boxes are observed. Regions currently aren't auto generated and thus there would need to be seperate observers attached to the seperate elements.

Now if we assume a regions that does auto generate a new box, similiar to how multi-col works where it clones it's sibling to generate a new box to contain the new content, it is a completely new element although matching the definition of a fragment. In this scenario I would expect that the Observer would be then attached to each additional fragment and its associated box to observe to ensure that any time that any box's dimensions change the callback is called and the above ResizeObserverEntry is passed.

There may be issues here with queueing up numerous additional Observers, some open questions I have is:
* Are dynamically clonable observers supported by Observer's currently
* What are the performance implications within the adjusted ResizeObserver lifecycle of queueing up potentially infinite Observers? I have a feeling we'll run into a situation where it will result in an increase of skipped observations and thus lead to increase of LoopErrors.
* @atotic it would valuable, with your work on LayoutNG to put together a rough dev assessment of supporting fragments in resizeObserver in conjuction with LayoutNG since it seems Firefox will be able to support this.

@plinss Primarily, does the ResizeObserverEntry object above address your concerns with support of fragments and this API not blocking the capality of using RO in conjuction for fragment based scenarios in the future?

Sorry for the book, just trying to think out loud here about the potential issues and the usecases where this may be utilized as I agree I don't want RO to become useless if we light up future scenarios where more fragmentation occurs.

I look forward to hearing your thoughts @atotic & @plinss

**Comment by atotic** [comment]
Re: Implementing fragments in Chromium

Our current layout engine support for fragments is a giant hack. Our new layout engine, LayoutNG, supports fragments natively. LayoutNG will be able to support fragment size observation. 

I'd rather not tackle fragment support in Legacy. We are hoping to ship LayoutNG fragment support within a year.

**Comment by atotic** [comment]
Re: reporting fragment size as array of sizes

Current proposal: object
`borderSize: { inline: ,block: }`
mutates to an array if target becomes fragmented
`borderSize: [ { inline: ,block: }`

My concern is this: 

If webdevs wrote their RO handler without taking fragmentation into account like this:

    new ResizeObserver(entries => {
      let newBorderWidth = entries[0].borderSize.inline / 10;
    },

their code would break if borderSize became an array.

I believe most webdevs will not write fragmentation-compatible code.

To minimize breakage of fragmented components that use ResizeObserver, I propose that developers should explicitly opt-in to fragment support. This can be done by adding a new option to RO.observer `fragments: false/true.`. If fragments is true, sizes would always be reported as an array.

**Comment by gregwhitworth** [comment]
@atotic `borderSize` is not currently shipping anywhere - as such, when authors learn about this change to the API they will undoubtedly either through tutorial or debugging see that it returns an array of entries. And thus their code will be written with `borderSize[0]`.

**Comment by gregwhitworth** [comment]
Another question, do we want to surface the `contentRect` dimensions in total alongside the fragments that contribute to these dimensions. I think we should as there may be information that are not surfaced in the fragments (eg: column gaps).

```
contentBox: {
           inlineSize: <double>
           blockSize: <double>
           fragments: [{ inlineSize: <double>, blockSize: <double> }]
}
```

**Comment by gregwhitworth** [comment]
To summarize the above - the decision was to go with Option 1 which is:

```
{
/* ... rest of object ... */
contentBox: [
        {
           inlineSize: <double>,
           blockSize: <double>
        },
        {
           inlineSize: <double>,
           blockSize: <double>
        }
]
}
```

In addition it was to recommended to add a visual example of a fragmented example and how to combine the fragments to be something useful (I'll open a new issue for this).

**Comment by dlibby-** [comment]
Addressed with #4529

**Comment by Loirooriol** [comment]
#4529 made it forwards compatible with fragment support, but didn't add fragment support, so reopening.
